auv_navigation/auv_navigation/scripts/follow_path_action_server.py

```diff
@@ -0,0 +1,238 @@
+from actionlib import SimpleActionServer
+import rospy
+import actionlib
+import numpy as np
+from nav_msgs.msg import Path
+from geometry_msgs.msg import Twist
+from std_msgs.msg import Bool
+from auv_msgs.msg import FollowPathAction, FollowPathFeedback, FollowPathResult, FollowPathGoal
+from auv_msgs.srv import AlignFrame, AlignFrameResponse
+import tf2_ros
+
+from navigation_utils.path_utils import (
+    create_path_from_frame, get_current_pose, 
+    calculate_carrot_pose, broadcast_carrot_frame
+)
+from navigation_utils.align_frame_utils import AlignFrameController
+
+class FollowPathActionServer:
+    def __init__(self):
+        rospy.init_node('follow_path_action_server')
+        self.tf_buffer = tf2_ros.Buffer()
+        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
+        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
+        self.carrot_distance = rospy.get_param('~carrot_distance', 1.0)
+        self.position_threshold = rospy.get_param('~position_threshold', 0.1)
+        self.angle_threshold = rospy.get_param('~angle_threshold', 0.1)
+        self.source_frame = rospy.get_param('~source_frame', "taluy/base_link")
+        self.carrot_frame = rospy.get_param('~carrot_frame', "carrot")
+        self.control_rate = rospy.Rate(rospy.get_param('~control_rate', 20))
+        self.frame_controller = AlignFrameController(
+            max_linear_velocity=rospy.get_param('~max_linear_velocity', 0.8),
+            max_angular_velocity=rospy.get_param('~max_angular_velocity', 0.9),
+            kp=rospy.get_param('~kp', 0.55),
+            angle_kp=rospy.get_param('~angle_kp', 0.45)
+        )
+
+        self.path_pub = rospy.Publisher('path', Path, queue_size=1)
+        self.cmd_vel_pub = rospy.Publisher('taluy/cmd_vel', Twist, queue_size=1) #TODO change to cmd_vel after given a namespace in start.launch
+        
+        self.killswitch_sub = rospy.Subscriber(
+            "/taluy/propulsion_board/status",
+            Bool,
+            self.killswitch_callback,
+        )
+        
+        rospy.Service(
+            "align_frame",
+            AlignFrame,
+            self.handle_align_frame_request
+        )
+        rospy.logdebug("AlignFrame Service is ready.")
+        
+        self.action_client = actionlib.SimpleActionClient(
+            "follow_path", FollowPathAction)
+
+        
+        self.server = actionlib.SimpleActionServer(
+            'follow_path',
+            FollowPathAction,
+            self.execute,
+            auto_start=False
+        )
+        self.server.start()
+        rospy.logdebug("FollowPath Action Server is ready.")
+        self.action_client.wait_for_server()
+        rospy.logdebug("Internal Action Client connected to the FollowPath Action Server.")
+        
+    def killswitch_callback(self, msg):
+        if not msg.data:
+            if self.server.is_active():
+                self.action_client.cancel_goal() 
+                rospy.logwarn("Control canceled")
```

Comment:
This function will cancel the current goal and cause the robot to stop. Notice how we don't need to have a service to cancel the action as in legacy implementation of align frame controller. We can simply
```
rostopic pub /follow_path/cancel actionlib_msgs/GoalID "{}"
```
Action server publishes these topics:
/follow_path/cancel
/follow_path/feedback
/follow_path/goal
/follow_path/result
/follow_path/status